server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -171,6 +172,44 @@ public DataStream promoteDataStream() {
         return new DataStream(name, timeStampField, indices, getGeneration(), metadata, hidden, false);
     }
 
+    /**
+     * Reconciles this data stream with its state at the end of the snapshot.
+     *
+     * @param afterSnapshot The post-snapshot data stream
+     * @return Reconciled {@link DataStream} instance
+     */
+    public DataStream reconcile(DataStream afterSnapshot) {
```

Comment:
The logic in this method makes sense to me. Shouldn't it be used by the snapshot service?

Reply:
Yes, @original-brownbear will incorporate into one of his PRs.

Reply:
>I think we shouldn't reject, but just use the indices' provided by afterSnapshot?

I don't think the indices from `afterSnapshot` will be present in the snapshot because they were created after the start of the snapshot.

Reply:
Ah right, sorry, that doesn't make sense. I think we shouldn't fail, because it is likely that some data streams will rollover while taking a snapshot. This would mean a different index would be the write index, the question is whether that would be a problem. Unless that index went read-only (e.g. due ilm managing the data stream) then writing to the data stream would fail, but then a rollover on the restored data stream would fix that.